Salt stateless CSRF tokens as well

markstory opened the pull request:
Add salting to CSRF tokens to prevent BREACH weakness. This ports the
changes made to session based CSRF tokens to stateless ones as well.
This will result in longer tokens as both the cookie and the POST data
are base64 encoded, and the POST data is now 112 bytes instead of 56
bytes.

markstory committed:
Salt stateless CSRF tokens as well

Add salting to CSRF tokens to prevent BREACH weakness. This ports the
changes made to session based CSRF tokens to stateless ones as well.
This will result in longer tokens as both the cookie and the POST data
are base64 encoded, and the POST data is now 112 bytes instead of 56
bytes.

othercorey commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -240,12 +304,18 @@ public function createToken(): string
+        // If base64 fails we're in a compatibility mode from before
+        // tokens were salted on each request.
+        if ($decoded === false) {
+            $decoded = $token;
```
Should we add a warning here? Users shouldn't see this occur more than once per request/refresh, yes?

ADmad replied:
I don't think we should show explicit warnings for internal changes which have been mitigated and don't require the dev to make any changes.

othercorey replied:
What I am asking is whether this introduces a security risk with the fallback.

markstory replied:
I don't think there is anything new being introduced. Without this flow we wouldn't be able to accept old tokens which would break applications as they upgrade.

othercorey replied:
Ok. As long as they ensure the token is refreshed. Up to them. We should probably test this though since codecov says it's untested.

othercorey commented:
> Strict comparison using === between string and false will always evaluate to false.

This seems like a bug in phpstan. We want it to return false for string comparisons, and it should since the return type is `string|false`.

ADmad commented:
> This seems like a bug in phpstan.

Similar code does not throw any error on it's playground https://psalm.dev/r/510895c6c3

othercorey commented:
The error is in phpstan: https://phpstan.org/r/d51986a7-1515-4135-8c1e-f872ac58f2e3

ADmad commented on src/Http/Middleware/CsrfProtectionMiddleware.php:
```diff
@@ -227,9 +240,60 @@ protected function _createToken(): string
+    public function unsaltToken(string $token): string
+    {
+        $decoded = base64_decode($token);
+        if ($decoded === false || strlen($decoded) !== static::TOKEN_WITH_CHECKSUM_LENGTH * 2) {
```
`base64_decode()` returns false only if 2nd argument is `true` https://www.php.net/manual/en/function.base64-decode.php

othercorey replied:
Good catch. Looks like they have a special case somewhere other than the functionMap I checked. No error when 2nd argument is true.

markstory replied:
Thanks for assist on this :bow:. I added the strict parameter to all the places we decode tokens from external sources.

markstory committed:
Add strict mode parameter to base64_encode

Decode operations on external data should use strict mode to reduce the
chances of decoding partially bad data.
